qa/rpc-tests/addressindex.py

```diff
@@ -0,0 +1,389 @@
+            'start': 1,
+            'end': 5
+        })
+        assert_equal(sorted(height_txids), sorted(unspent_txids))
+
+        # each txid should appear only once
+        height_txids = self.nodes[1].getaddresstxids({
+            'addresses': [addr_p2pkh, addr_p2sh],
+            'start': 1,
+            'end': 5
+        })
+        assert_equal(sorted(height_txids), sorted(unspent_txids))
+
+        # do some transfers, make sure balances are good
+        txids_a = []
+        a = self.nodes[1].getnewaddress()
+        expected = 0
+        expected_deltas = [] # for checking getaddressdeltas (below)
+        for i in range(5):
+            # first transaction happens at height 105, mined in block 106
+            txid = self.nodes[0].sendtoaddress(a, i+1)
+            txids_a.append(txid)
+            self.nodes[0].generate(1)
+            self.sync_all()
+            expected += i+1
+            expected_deltas.append({
+                'height': 106+i,
+                'satoshis': (i+1)*COIN,
+                'txid': txid,
+            })
+        bal = self.nodes[1].getaddressbalance(a)
+        assert_equal(bal['balance'], expected*COIN)
+        assert_equal(bal['received'], expected*COIN)
+        assert_equal(sorted(self.nodes[0].getaddresstxids(a)), sorted(txids_a))
+        assert_equal(sorted(self.nodes[1].getaddresstxids(a)), sorted(txids_a))
+
+        # Restart all nodes to ensure indices are saved to disk and recovered
+        stop_nodes(self.nodes)
+        wait_bitcoinds()
+        self.setup_network()
+
+        bal = self.nodes[1].getaddressbalance(a)
+        assert_equal(bal['balance'], expected*COIN)
+        assert_equal(bal['received'], expected*COIN)
+        assert_equal(sorted(self.nodes[0].getaddresstxids(a)), sorted(txids_a))
+        assert_equal(sorted(self.nodes[1].getaddresstxids(a)), sorted(txids_a))
+
+        # Send 3 from address a, but -- subtlety alert! -- address a at this
+        # time has 4 UTXOs, with values 1, 2, 3, 4. Sending value 3 requires
+        # using up the value 4 UTXO, because of the tx fee
+        # (the 3 UTXO isn't quite large enough).
+        #
+        # The txid from sending *from* address a is also added to the list of
+        # txids associated with that address (test will verify below).
+
+        b = self.nodes[2].getnewaddress()
+        txid = self.nodes[1].sendtoaddress(b, 3)
+        self.sync_all()
+
+        # the one tx in the mempool refers to addresses a and b
+        mempool = self.nodes[0].getaddressmempool({'addresses': [b, a, b]})
+        assert_equal(len(mempool), 3)
+        assert_equal(mempool[1]['address'], a)
+        assert_equal(mempool[1]['satoshis'], (-4)*COIN)
+        assert_equal(mempool[1]['txid'], txid)
+        for i in (0, 2):    # address b listed in args first and last
+            assert_equal(mempool[i]['address'], b)
+            assert_equal(mempool[i]['satoshis'], 3*COIN)
+            assert_equal(mempool[i]['txid'], txid)
```

Comment:
This could become:
```
        mempool = self.nodes[0].getaddressmempool({'addresses': [addr2, addr1, addr2]})
        assert_equal(len(mempool), 3)
        [assert_equal(tx['txid'], txid) for tx in mempool]
        # addr2 listed in args first and last
        assert_equal(mempool[0]['address'], addr2)
        assert_equal(mempool[0]['satoshis'], 3 * COIN)
        assert_equal(mempool[1]['address'], addr1)
        assert_equal(mempool[1]['satoshis'], (-4) * COIN)
        assert_equal(mempool[2]['address'], addr2)
        assert_equal(mempool[2]['satoshis'], 3 * COIN)
```

If we wanted to eliminate the loop (which I found marginally confusing).